File: src/e111_minimum_depth_of_binary_tree.rs
// https://leetcode.cn/problems/minimum-depth-of-binary-tree/
// 给定一个二叉树，找出其最小深度。
// 最小深度是从根节点到最近叶子节点的最短路径上的节点数量。

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}
//
impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
use std::cell::RefCell;
use std::cmp::{max, min};
use std::rc::Rc;

struct Solution {}
impl Solution {
    pub fn min_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        Self::min_depth_node(root.as_ref())
    }

    fn min_depth_node(node: Option<&Rc<RefCell<TreeNode>>>) -> i32 {
        match node {
            None => 0,
            Some(node) => {
                let left_depth = Self::min_depth_node(node.borrow().left.as_ref());
                let right_depth = Self::min_depth_node(node.borrow().right.as_ref());
                match (left_depth, right_depth) {
                    (0, _) => right_depth + 1,
                    (_, 0) => left_depth + 1,
                    _ => min(left_depth, right_depth) + 1,
                }
            }
        }
    }
}

File: javascript/src/e111_minimum_depth_of_binary_tree.js
// 111. 二叉树的最小深度
// https://leetcode.cn/problems/minimum-depth-of-binary-tree/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function (root) {
  if (root) {
    if (root.left && root.right) {
      return Math.min(minDepth(root.left), minDepth(root.right)) + 1
    }
    if (root.left && !root.right) {
      return minDepth(root.left) + 1;
    }
    if (!root.left && root.right) {
      return minDepth(root.right) + 1;
    }
    if (!root.left && !root.right) {
      return 1;
    }
  } else {
    return 0;
  }
};